### ISSUE TITLE
`torch.compile` silently computes wrong gradients for chained modulated LayerNorms with a shared scale/shift parameter (regression 2.9.1 → 2.10.0 due to mix_order_reduction)

### ISSUE THREAD

**Issue by Lakonik** [pytorch/pytorch#193061]
### 🐛 Describe the bug

When a model applies two chained "modulated norm" blocks of the form `x = x + (layer_norm(x) * (1 + scale) + shift)`, where all `scale`/`shift` vectors are **chunks of one shared parameter** (the DiT / AdaLN `scale_shift_table` pattern used by Wan 2.1), the Inductor-compiled backward produces **silently wrong gradients for the second norm's `shift`/`scale` slices** once the total reduction size `batch*seq` is large (≳16k tokens). The first norm's slices are correct to fp32 precision; the second norm's slices come back with relative errors of up to ~1e+6 — depending on graph arrangement the corresponding region of the grad buffer is simply never written, so it can also surface as bitwise-stale values from a previous backward.

**Update: This is a regression introduced in 2.10.0**: the same script is fully clean on 2.9.1 and fails identically on 2.10.0, 2.11.0, and 2.13.0 (latest stable).

Minimal repro in plain fp32, plain `nn.LayerNorm`, no autocast, no checkpointing, no attention:

```python
import torch
import torch.nn as nn

DIM, SEQ, SMALL_SEQ = 1024, 32768, 4096
dev = 'cuda'


def run(seq=SEQ, n_norms=2, slicing='chunk'):
    torch.manual_seed(0)
    norms = [nn.LayerNorm(DIM, 1e-6, elementwise_affine=False).to(dev)
             for _ in range(n_norms)]
    if slicing == 'separate':
        params = [nn.Parameter(torch.randn(DIM, device=dev) / DIM ** 0.5)
                  for _ in range(2 * n_norms)]
    else:
        table = nn.Parameter(
            torch.randn(1, 2 * n_norms, DIM, device=dev) / DIM ** 0.5)
        params = [table]

    def fn(x):
        if slicing == 'separate':
            mods = params
        elif slicing == 'unbind':
            mods = [s.unsqueeze(1) for s in params[0].unbind(1)]
        else:  # 'chunk' — the failing case
            mods = params[0].chunk(2 * n_norms, dim=1)
        for i, norm in enumerate(norms):
            shift, scale = mods[2 * i], mods[2 * i + 1]
            x = x + (norm(x) * (1 + scale) + shift)
        return x.square().mean()

    fn_c = torch.compile(fn, mode='default', dynamic=False)
    x = torch.randn(1, seq, DIM, device=dev)

    def grads():
        g = [p.grad for p in params]
        for p in params:
            p.grad = None
        return torch.cat([t.reshape(-1) for t in g])

    fn(x).backward()
    g_eager = grads()
    fn_c(x).backward()
    g_compiled = grads()

    rel = ((g_eager - g_compiled).abs().max()
           / g_eager.abs().max().clamp(min=1e-12)).item()
    return rel


CASES = [
    ('chunked table, 2 norms, seq 32768  (THE BUG)', dict()),
    ('same but seq 4096', dict(seq=SMALL_SEQ)),
    ('same but single norm', dict(n_norms=1)),
    ('same but unbind(1)+unsqueeze(1) instead of chunk', dict(slicing='unbind')),
    ('same but 4 independent parameters', dict(slicing='separate')),
]

for name, kwargs in CASES:
    rel = run(**kwargs)
    print(f'{"FAIL" if rel > 0.1 else "ok  "}  max rel grad err {rel:9.3e}  {name}',
          flush=True)

print('\nexpected: first row FAIL (silently wrong grads), all others ok')
```

### Output

On H100 — **identical on torch 2.10.0, 2.11.0 and 2.13.0** (cu126/cu128/cu129 builds all affected; CUDA toolkit version is irrelevant). **torch 2.9.1 is clean** (first row `ok  2.141e-07`):

```
FAIL  max rel grad err 1.481e+06  chunked table, 2 norms, seq 32768  (THE BUG)
ok    max rel grad err 2.068e-07  same but seq 4096
ok    max rel grad err 2.251e-07  same but single norm
ok    max rel grad err 2.141e-07  same but unbind(1)+unsqueeze(1) instead of chunk
ok    max rel grad err 2.116e-07  same but 4 independent parameters
```

### What we bisected (each ingredient is necessary)

1. **Two chained modulated norms** — a single one is clean.
2. **Gradient flow through a multiplicative rsqrt-variance statistic** — an RMSNorm-style `x * rsqrt(mean(x^2) + eps)` fails exactly like LayerNorm; removing the norm, subtracting only the mean (`x - x.mean(-1, keepdim=True)`), or detaching the norm statistics (same forward, pointwise backward) are all clean.
3. **The modulation vectors must be `chunk`/narrow slices of one shared tensor** — `unbind(1)` (+`unsqueeze(1)` for the broadcast shape) or independent parameters are clean. The shared tensor need not be a parameter: chunking a conditioning MLP's output fails the same way (max rel err 7.2e+1 / 4.7e+5 on 2.11 / 2.13, landing in the MLP weight grads). In every failing case the slice grads assemble into one buffer — matching the never-written/stale-buffer behavior.
4. **The slices must broadcast over batch AND sequence, with total `batch*seq` ≳ 16k** — at DIM=5120: seq 8192 ok, seq 16384 FAIL; also FAIL at DIM=256 with long seq, and at batch 8 × seq 4096 (shared table) even though batch 1 × seq 4096 is ok. Per-sample modulation (a real batch dim, e.g. `lin(c).view(B, 4, D)` with B ≥ 2) is clean at every tested length (B=2 × seq 32768, B=4 × seq 65536 ok) — but at B=1 it degenerates to the broadcast case and fails (B=1 × seq 32768 FAIL). Note the two roles of batch size: within the broadcast case it only contributes to the total (B=8 × 4096 fails like B=1 × 32768), while B ≥ 2 per-sample modulation is clean even with a 65536-long reduction per grad element — so the broadcast structure selects the buggy codegen path, and reduction size is the trigger only within it.

Not required: autocast/bf16, gradient checkpointing, attention/FFN, `elementwise_affine`. No escape via compile config (`default`/`reduce-overhead`/`max-autotune`, `fullgraph` on/off, `dynamic` on/off), `TORCHINDUCTOR_SPLIT_REDUCTIONS=0`, or expression rewrites (cloning the norm output or slices, `contiguous()`, `torch.addcmul`, distributing the multiply) — only the `unbind`/separate-parameter rewrites, which change how the slice grads are routed, avoid it.

### Real-world impact

Found while training diffusers Wan 2.1 14B (video diffusion): the per-block `scale_shift_table` gradients were garbage, and since the grad-buffer region is never written, iteration N reused iteration N-1's values bitwise whenever shapes matched — a linearly growing grad-norm staircase.

Who is exposed: any chained-modulated-norm model whose modulation slices broadcast over the batch — learnable scale-shift tables (Wan-style, at any batch size), or ordinary per-sample adaLN whenever the local batch is 1 (common in long-video training and gradient accumulation). Per-sample adaLN at local batch ≥ 2 tested clean at every length (up to seq 65536), and we verified three real DiT-family ImageNet training configs are unaffected.

### Versions

Fails on 2.10.0+cu128, 2.11.0+cu128/cu129, 2.13.0+cu126/cu129; clean on 2.9.1+cu128 (regression window: 2.9.1 → 2.10.0). `collect_env` for the primary environment:

```
PyTorch version: 2.13.0+cu129
Is debug build: False
CUDA used to build PyTorch: 12.9
OS: Ubuntu 22.04.5 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04.3) 11.4.0
Libc version: glibc-2.35
Python version: 3.11.10 (64-bit runtime)
Python platform: Linux-5.15.0-131-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 12.9.86
GPU models and configuration: 8x NVIDIA H100 80GB HBM3
Nvidia driver version: 550.127.08
cuDNN version: 9.10.2
triton: 3.7.1
```

### A related observation

On 2.11.0 only, the same setup with independent parameters of shape `[1, 1, D]` (instead of `[D]` used above) made the compiled backward crash outright (`Function CompiledFunctionBackward returned an invalid gradient at index 3 - got [1024] but expected shape compatible with [1, 1, 1024]`). 2.13.0 neither crashes nor miscomputes there — so the grad-shape handling in this area changed between 2.11 and 2.13 while the silent bug above survived; possibly a useful hint for locating it.

### AI assistance disclosure

Per the [PyTorch AI policy](https://github.com/pytorch/pytorch/blob/main/AI_POLICY.md):
The bug was hit in a real training run (Wan 2.1 14B). The debugging, ablation, and this write-up were developed with AI assistance (Claude Fable 5). Every number and code path in this report was executed and verified on our cluster (H100, both torch versions), and I have reviewed and can speak to all of it.

cc @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @aakhundov @coconutruben @jataylo @ezyang @gchanan @zou3519 @msaroufim @eellison @shunting314 @jansel @bdhirsh

**Comment by Lakonik** [comment]
Update: bisected the version range — **this is a regression introduced in 2.10.0**. torch 2.9.1 is fully clean. 2.10.0 fails with the identical signature as 2.11/2.13.

Running the exact script from the issue body:

```
# torch 2.9.1+cu128 — clean
ok    max rel grad err 2.141e-07  chunked table, 2 norms, seq 32768  (THE BUG)
ok    max rel grad err 1.378e-07  same but seq 4096
ok    max rel grad err 2.251e-07  same but single norm
ok    max rel grad err 2.141e-07  same but unbind(1)+unsqueeze(1) instead of chunk
ok    max rel grad err 2.116e-07  same but 4 independent parameters

# torch 2.10.0+cu128 — fails like 2.11/2.13
FAIL  max rel grad err 1.481e+06  chunked table, 2 norms, seq 32768  (THE BUG)
ok    max rel grad err 2.068e-07  same but seq 4096
ok    max rel grad err 2.251e-07  same but single norm
ok    max rel grad err 2.141e-07  same but unbind(1)+unsqueeze(1) instead of chunk
ok    max rel grad err 2.116e-07  same but 4 independent parameters
```

Full verified matrix: clean on 2.9.1+cu128; fails on 2.10.0+cu128, 2.11.0+cu128/cu129, 2.13.0+cu126/cu129 (same H100 machine). I've updated the issue body accordingly.

**Comment by primecai** [comment]
Root cause may be the **mix-order reduction** path, present and default-enabled in OSS from
2.10: `torch._inductor.config.triton.mix_order_reduction` is absent in 2.9.1 and defaults to
on in 2.10.0, matching the observed regression window.

Workaround, with `torch.compile` still enabled:

```bash
TORCHINDUCTOR_MIX_ORDER_REDUCTION=0
```

Verified to restore eager agreement (max relative error ≈ 2e-7) on 2.10.0 and 2.11.0 / H200 at
seq=32768/dim=1024, seq=16384/dim=5120 and batch=8/seq=4096/dim=1024.

In the generated backward, the four gradient slices of the `[1, 4, D]` modulation table should
each be a `reinterpret_tensor` view into one buffer. 2.9.1 emits all four; 2.10.0 emits only
the first two, and the missing two are exactly the second norm's `c_shift` / `c_scale`. In the
mix-order path those two are materialized as standalone tensors out of the reduction
workspace and have no consumer, so their regions of the returned `empty_strided` buffer keep
unspecified contents.

**Comment by haojiang01** [comment]
Thanks for the excellent diagnosis — the bisect to 2.10 and the observation that the second norm's slices are "materialized as standalone tensors out of the reduction workspace with no consumer" pointed straight at the root cause.

Confirmed: in `_codegen_mix_order_reduction`, the final-reduce step emits `buf = workspace[...].sum(dim=0)`, which rebinds the Python variable. When the reduction output is a view into a shared buffer (`NonOwningLayout`, here the `cat` assembling the chunked table's grad slices), the `reinterpret_tensor` alias is discarded and that region of the grad buffer is never written — hence the uninitialized/bitwise-stale values.

Fix in #193103: write through the view with `copy_()` instead of rebinding. Verified against the repro here plus the batch-8, dim-5120, and chunked-MLP-output ablations. Until it lands, `TORCHINDUCTOR_MIX_ORDER_REDUCTION=0` remains the correct workaround.
